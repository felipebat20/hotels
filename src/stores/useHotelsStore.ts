import { defineStore } from 'pinia';

import type { HotelWithPlaces, HotelInterface } from 'src/interfaces/HotelInterface';

import hotels_data from 'src/data/hotel.json';


export const useHotelsStore = defineStore('hotels', {
  state: () => ({
    hotels: [] as HotelInterface[],
    control_flow: {
      sortBy: '',
      name: '',
      placeId: '',
      request_pending: false,
      page: 1,
      has_next_page: true,
    },
  }),

  getters: { getHotels: (state) => state.hotels, },
  actions: {
    async fetchHotels({
      placeId,
      sortBy,
      page = 1,
      name
    }: {
        placeId?: number | string
        page?: number
        name?: string
        sortBy?: string
    }) {
      const all_hotels = [...hotels_data] as HotelWithPlaces[];

      try {
        let filtered_hotels = all_hotels.find((hotel) => hotel.placeId === placeId)?.hotels ||
          all_hotels.flatMap((hotel) => hotel.hotels) || [];


        if (! filtered_hotels.length) {
          this.control_flow.has_next_page = false;

          return this.hotels = [];
        }

        if (name) {
          filtered_hotels = filtered_hotels.filter((hotel) =>
            hotel.name.toLowerCase().includes(name.toLowerCase()),
          )
        }

        if (sortBy === 'best_rated') {
          filtered_hotels = filtered_hotels.sort((a, b) => +b.stars - +a.stars);
        }

        const paginated_hotels = filtered_hotels.slice((page - 1) * 10, page * 10);

        await new Promise((resolve) => setTimeout(resolve, 1000));

        this.control_flow.has_next_page = paginated_hotels.length >= 10;
        if (page === 1) {
          return this.hotels = paginated_hotels;
        }

        return this.hotels = [...this.hotels, ...paginated_hotels] as HotelInterface[];
      } catch (error) {
        return error;
      }
    }
  },
})