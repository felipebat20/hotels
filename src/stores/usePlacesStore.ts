import { defineStore } from 'pinia'

import { Place } from 'src/models/Place'
import type { PlaceInterface } from 'src/interfaces/PlaceInterface'

import placesMock from 'src/data/place.json'

const places = placesMock as PlaceInterface[]

export const usePlacesStore = defineStore('places', {
  state: () => ({ places: [] as Place[] }),
  actions: {
    async fetchPlaces() {
      try {
        await new Promise((resolve) => setTimeout(resolve, 1000))

        this.places = places.map((place) => new Place(place))
      } catch (error: unknown) {
        return error
      }
    },
  },
})