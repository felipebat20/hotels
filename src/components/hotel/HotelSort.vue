<template>
  <div class="row items-center">
    <p class="q-mb-none text-grey-7">
      Organizar por
    </p>

    <q-select
      v-model="selected_sort"
      :options="sortOptions"
      dense
      borderless
      options-dense
      @update:model-value="handleSort"
    >
      <template v-slot:selected>
        <span class="q-mb-none q-ml-xs text-blue text-weight-medium text-italic">
          {{ selected_sort.label }}
        </span>
      </template>
    </q-select>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import type { Ref } from 'vue';

import { useHotelsStore } from 'src/stores/useHotelsStore';

interface SortOption {
  label: string;
  value: string;
}

const emit = defineEmits(['fetch-hotels'])

const sortOptions: Ref<SortOption[]> = ref([
  {
    label: 'Recomendados',
    value: 'recommended'
  },

  {
    label: 'Melhor avaliados',
    value: 'best_rated'
  },
]);

const selected_sort: Ref<SortOption> = ref(sortOptions.value[0] as SortOption);
const hotelsStore = useHotelsStore();

const handleSort = async ({ value }: { value: string }) => {
  hotelsStore.control_flow.sortBy = value;
  hotelsStore.hotels = [];
  hotelsStore.control_flow.request_pending = true;

  await hotelsStore.fetchHotels({
    sortBy: value,
    name: hotelsStore.control_flow.name,
    placeId: hotelsStore.control_flow.placeId,
    page: 1,
  });

  hotelsStore.control_flow.request_pending = false;
};
</script>