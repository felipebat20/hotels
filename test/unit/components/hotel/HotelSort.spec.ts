import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest';
import { mount } from '@vue/test-utils';
import { describe, expect, it, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing'

import HotelSort from '@/components/hotel/HotelSort.vue';
import { useHotelsStore } from 'src/stores/useHotelsStore';

installQuasarPlugin();

describe('HotelSort', () => {
  const createWrapper = () => {
    const wrapper = mount(HotelSort, {
      global: {
        plugins: [createTestingPinia({ createSpy: vi.fn })],
      },
    });

    const store = useHotelsStore();
    return { wrapper, store };
  };

  it('should render component properly', () => {
    const { wrapper } = createWrapper();
    const html = wrapper
      .html()
      .replace(/id="f_[^"]+"/g, 'id="static-id"')
      .replace(/for="f_[^"]+"/g, 'for="static-for"')
      .replace(/aria-controls="f_[^"]+"/g, 'aria-controls="static-aria-controls"');

    expect(html).toMatchSnapshot();
  });

  it('should display "Recomendados" as default sort option', () => {
    const { wrapper } = createWrapper();
    const selectedSort = wrapper.find('.text-blue');

    expect(selectedSort.text()).toBe('Recomendados');
  });

  it('should call fetchHotels when sort option changes', async () => {
    const { wrapper, store } = createWrapper();
    await wrapper.findComponent({ name: 'q-select' }).vm.$emit('update:modelValue', { value: 'best_rated' });

    expect(store.fetchHotels).toHaveBeenCalledTimes(1);
  });

  it('should clear hotels array when changing sort', async () => {
    const { wrapper, store } = createWrapper();
    await wrapper.findComponent({ name: 'q-select' }).vm.$emit('update:modelValue', { value: 'best_rated' });

    expect(store.hotels).toEqual([]);
  });
});
