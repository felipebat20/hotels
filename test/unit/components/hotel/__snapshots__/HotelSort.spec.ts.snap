// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelSort > should render component properly 1`] = `
"<div class="row items-center">
  <p class="q-mb-none text-grey-7"> Organizar por </p><label class="q-field row no-wrap items-start q-field--borderless q-select q-field--auto-height q-select--without-input q-select--without-chips q-select--single q-field--float q-field--dense" for="static-for">
    <!---->
    <div class="q-field__inner relative-position col self-stretch">
      <div class="q-field__control relative-position row no-wrap" tabindex="-1">
        <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip">
          <div class="q-field__native row items-center"><span class="q-mb-none q-ml-xs text-blue text-weight-medium text-italic">Recomendados</span><input class="q-select__focus-target" id="static-id" readonly="" tabindex="0" role="combobox" aria-readonly="false" aria-autocomplete="none" aria-expanded="false" aria-controls="static-aria-controls" value="Recomendados"></div>
          <!---->
        </div>
        <div class="q-field__append q-field__marginal row no-wrap items-center q-anchor--skip"><i class="q-icon notranslate material-icons q-select__dropdown-icon" aria-hidden="true">arrow_drop_down</i></div>
        <!---->
      </div>
      <!---->
    </div>
    <!---->
  </label>
</div>"
`;
